v3/integration/lints/lint/lint.go

```diff
@@ -0,0 +1,139 @@
+package lint
+
+/*
+ * ZLint Copyright 2021 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+)
+
+type Lint interface {
+	Lint(tree *ast.File, file *File) *Result
+	CheckApplies(tree *ast.File, file *File) bool
+}
+
+type Result struct {
+	message       string
+	codeCitations []string
+	citations     []string
+}
+
+func NewResult(message string) *Result {
+	return &Result{message: message}
+
+}
+
+func (r *Result) AddCodeCitation(start, end token.Pos, file *File) *Result {
+	srcCode := make([]byte, end-start)
+	reader := strings.NewReader(file.Src)
+	reader.ReadAt(srcCode, int64(start))
+	lineno := file.LineOf(start)
+	citation := fmt.Sprintf("File %s, line %d\n\n%s\n\n", file.Path, lineno, string(srcCode))
+	r.codeCitations = append(r.codeCitations, citation)
+	return r
+}
+
+func (r *Result) SetCitations(citations ...string) *Result {
+	r.citations = citations
+	return r
+}
+
+func (r *Result) String() string {
```

Comment:
This seems reasonable as-is, but if it gets any more complex I think it would be a good candidate to refactor with https://golang.org/pkg/text/template/